doc/specs/#2871 - Pane Navigation/#2871 - Pane Navigation.md

```diff
@@ -0,0 +1,298 @@
+
+Previously, this design was avoided, because what does `focusPane(target=4,
+direction=down)` do? Does it focus pane 4, or does it move focus down?
+
+`tmux` solves this in one action by just doing both!
+
+```
+Make pane target-pane the active pane ...  If one of -D, -L, -R, or -U is used,
+respectively the pane below, to the left, to the right, or above the target pane
+is used.
+```
+_from `man tmux`_.
+
+So `focusPane(target=1, direction=up)` will attempt to focus the pane above pane
+1. This action would not summon the pane switcher UX, even for
+`focusPane(direction=last)`
+
+* **Pros**:
+  - Fewest redundant actions
+* **Cons**:
+  - Is this intuitive? That combining the params would do both, with `target`
+    happening "first"?
+  - Assumes that there will be a separate action added in the future for "Open
+    the pane switcher (with some given ordering)"
+
+
+> 👉 **NOTE**: At this point, the author considered "Do we even want a separate
+> action to engage the tab switcher with panes expanded?" Perhaps panes being
+> visible in the tab switcher is just part fo the tab switcher's behavior. Maybe
+> there shouldn't be a separate "open the tab switcher with the panes expanded
+> to the pane I'm currently on, and the panes listed in MRU order" action.
+
+## Conclusion
+
+After much discussion as a team, we decided that **Proposal D** would be the
+best option. We felt that there wasn't a need to add any extra configuration to
+invoke the "pane switcher" as anything different than the "tab switcher". The
+"pane switcher" should really just exist as a part of the functionality of the
+advanced tab switcher, not as it's own thing.
+
+Additionally, we concurred that the new "direction" value should be `prev`, not
+`last`, for consistency's sake.
+
+## UI/UX Design
+
+The only real UX being added with the agreed upon design is allowing the user to
+execute an action to move to the previously active pane within a single tab. No
+additional UX (including the pane switcher) is being prescribed in this spec at
+this time.
+
+## Potential Issues
+
+<table>
+<tr>
+<td><strong>Compatibility</strong></td>
+<td>
+
+We've only adding a single enum value to an existing enum. Since we're not
+changing the meaning of any of the existing values, we do not expect any
+compatibility issues there. Additionally, we're not changing the default value
+of the `direction` param of the `moveFocus` action, so there are no further
+compatibility concerns there. Furthermore, no additional parameters are being
+added to the `moveFocus` action that would potentially gibe it a different
+meaning.
+
+</td>
+</tr>
+</table>
+
+In the current design, there's no way to move through all the panes with a
+single keybinding. For example, if a user wanted to bind <kbd>Alt+]</kbd> to
+move to the "next" pane, and <kbd>Alt+[</kbd> to move to the "previous" one.
+These movements would necessarily need to be in-order traversals, since there's
+no way of doing multiple MRU steps.

```

Comment:
Is in-order traversal...
- order the panes were opened, or
- order the panes were interacted with?

I'd like a bit more discussion on the second one. I still don't understand the issue with the following approach:
- maintain a linked list of weak references to Panes
- prev/next navigates the list
- if a pane was removed, skip the node
- if a pane receives focus (except from prev/next action)...
  - clear all "next" panes from the list
  - replace with a weak ref to the newly focused pane

In my mind, that's how Ctrl+Z/Y works in text editors and how Back/Next work in web browsers.

Reply:
"in-order" is always "the order defined by the pane IDs" and "MRU order" is always the "order in which they were last interacted with".

I don't know if your comment was from before we discussed this in the meeting or not. TL;DR: Without another transient UI to manage the navigation of the MRU stack, it's not possible to know the difference between focus moving because of MRU pane navigation, or the user interacting with it. We'd have to hook all the possible interaction events for the control, and then manually handle them as "commit this navigation", and it's just a mess